File: frc2018_4/src/main/java/ca/warp7/frc2018_4/subsysterms/Climber.java
package ca.warp7.frc2018_4.subsysterms;

import ca.warp7.frc.core.ISubsystem;
import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
import edu.wpi.first.wpilibj.SpeedControllerGroup;

import static ca.warp7.frc.Functions.limit;
import static ca.warp7.frc.core.Robot.reportInputAndState;
import static ca.warp7.frc2018_4.constants.ClimberConstants.kAbsoluteMaxOutputPower;
import static ca.warp7.frc2018_4.constants.ClimberConstants.kRandomDiffDivision;
import static ca.warp7.frc2018_4.constants.Pins.kClimberPin;

public class Climber implements ISubsystem {

    private SpeedControllerGroup mClimbMotorGroup;
    public InputState mInputState = new InputState();
    public State mState = new State();
    @Override
    public void onConstruct() {
        mClimbMotorGroup = new SpeedControllerGroup(new WPI_VictorSPX(kClimberPin));
    }

    @Override
    public void onDisabled() {
        mInputState.mDemandedMotorSpeed = 0;
        mClimbMotorGroup.disable();
    }
    @Override
    public void onOutput() {
        mClimbMotorGroup.set(limit(mState.mSpeed, kAbsoluteMaxOutputPower));
    }

    @Override
    public void onUpdateState() {
        mState.mSpeed += (mState.mSpeed - mInputState.mDemandedMotorSpeed)/kRandomDiffDivision;
    }

    @Override
    public void onReportState() {
        reportInputAndState(this, mInputState, mState);
    }
    public class InputState{
        public double mDemandedMotorSpeed;
    }
    public class State{
        double mSpeed;
    }
}

File: frc2018_5/src/main/java/ca/warp7/frc2018_5/output/Climber.kt
package ca.warp7.frc2018_5.output

import ca.warp7.frc.next.OutputSystem
import ca.warp7.frc2018_5.constants.RobotPins.kClimberPin
import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.can.VictorSPX

object Climber : OutputSystem {
    private val climberMaster = VictorSPX(kClimberPin)

    var percentOutput = 0.0

    @Synchronized
    override fun onDisabled() {
        climberMaster.set(ControlMode.Disabled, 0.0)
    }

    override fun onIdle() {
        onDisabled()
    }

    @Synchronized
    override fun onOutput() {
        climberMaster.set(ControlMode.PercentOutput, percentOutput)
    }
}
